### DISCUSSION TITLE
libmesh installation error with petsc 3.19.0

### DISCUSSION THREAD

**Discussion by aeronova** [libMesh/libmesh#3511] [General]
Hello,

It looks petsc has changed some function names. I got the following error when I tried to install the git version of libmesh with petsc 3.19.0 and slepc 3.19.0.

In file included from src/solvers/petsc_linear_solver.C:30:
./include/libmesh/petsc_vector.h:738:44: warning: PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19) [-W#pragma-messages]
  PetscInt * petsc_ghost = ghost.empty() ? PETSC_NULL :
                                           ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscsys.h:195:22: note: expanded from macro 'PETSC_NULL'
#define PETSC_NULL   PETSC_DEPRECATED_MACRO("GCC warning \"PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)\"") PETSC_NULLPTR
                     ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscconf.h:12:37: note: expanded from macro 'PETSC_DEPRECATED_MACRO'
#define PETSC_DEPRECATED_MACRO(why) _Pragma(why)
                                    ^
<scratch space>:111:6: note: expanded from here
 GCC warning "PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)"
     ^
src/solvers/petsc_linear_solver.C:202:36: warning: PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19) [-W#pragma-messages]
                                   PETSC_NULL,   // pointer to the array which holds the history
                                   ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscsys.h:195:22: note: expanded from macro 'PETSC_NULL'
#define PETSC_NULL   PETSC_DEPRECATED_MACRO("GCC warning \"PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)\"") PETSC_NULLPTR
                     ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscconf.h:12:37: note: expanded from macro 'PETSC_DEPRECATED_MACRO'
#define PETSC_DEPRECATED_MACRO(why) _Pragma(why)
                                    ^
<scratch space>:136:6: note: expanded from here
 GCC warning "PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)"
     ^
src/solvers/petsc_linear_solver.C:288:36: warning: PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19) [-W#pragma-messages]
                                   PETSC_NULL,   // pointer to the array which holds the history
                                   ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscsys.h:195:22: note: expanded from macro 'PETSC_NULL'
#define PETSC_NULL   PETSC_DEPRECATED_MACRO("GCC warning \"PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)\"") PETSC_NULLPTR
                     ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscconf.h:12:37: note: expanded from macro 'PETSC_DEPRECATED_MACRO'
#define PETSC_DEPRECATED_MACRO(why) _Pragma(why)
                                    ^
<scratch space>:138:6: note: expanded from here
 GCC warning "PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)"
     ^
src/solvers/petsc_linear_solver.C:904:10: warning: 'KSP_CONVERGED_CG_NEG_CURVE' is deprecated: Use KSP_CONVERGED_NEG_CURVE (since version 3.19) [-Wdeprecated-declarations]
    case KSP_CONVERGED_CG_NEG_CURVE    : return CONVERGED_CG_NEG_CURVE;
         ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscksp.h:669:3: note: 'KSP_CONVERGED_CG_NEG_CURVE' has been explicitly marked deprecated here
  KSP_CONVERGED_CG_NEG_CURVE_DEPRECATED   = 5,
  ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscksp.h:622:76: note: expanded from macro 'KSP_CONVERGED_CG_NEG_CURVE_DEPRECATED'
#define KSP_CONVERGED_CG_NEG_CURVE_DEPRECATED   KSP_CONVERGED_CG_NEG_CURVE PETSC_DEPRECATED_ENUM("Use KSP_CONVERGED_NEG_CURVE (since version 3.19)")
                                                                           ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscconf.h:10:51: note: expanded from macro 'PETSC_DEPRECATED_ENUM'
#define PETSC_DEPRECATED_ENUM(why) __attribute__((deprecated(why)))
                                                  ^
src/solvers/petsc_linear_solver.C:905:10: warning: 'KSP_CONVERGED_CG_CONSTRAINED' is deprecated: Use KSP_CONVERGED_STEP_LENGTH (since version 3.19) [-Wdeprecated-declarations]
    case KSP_CONVERGED_CG_CONSTRAINED  : return CONVERGED_CG_CONSTRAINED;
         ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscksp.h:670:3: note: 'KSP_CONVERGED_CG_CONSTRAINED' has been explicitly marked deprecated here
  KSP_CONVERGED_CG_CONSTRAINED_DEPRECATED = 6,
  ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscksp.h:623:78: note: expanded from macro 'KSP_CONVERGED_CG_CONSTRAINED_DEPRECATED'
#define KSP_CONVERGED_CG_CONSTRAINED_DEPRECATED KSP_CONVERGED_CG_CONSTRAINED PETSC_DEPRECATED_ENUM("Use KSP_CONVERGED_STEP_LENGTH (since version 3.19)")
                                                                             ^
/Users/aeronova/Development/local/lib/petsc/3.19.0/include/petscconf.h:10:51: note: expanded from macro 'PETSC_DEPRECATED_ENUM'
#define PETSC_DEPRECATED_ENUM(why) __attribute__((deprecated(why)))
                                                  ^
src/solvers/petsc_linear_solver.C:906:10: error: duplicate case value: 'KSP_CONVERGED_CG_CONSTRAINED' and 'KSP_CONVERGED_STEP_LENGTH' both equal '6'
    case KSP_CONVERGED_STEP_LENGTH     : return CONVERGED_STEP_LENGTH;
         ^
src/solvers/petsc_linear_solver.C:905:10: note: previous case defined here
    case KSP_CONVERGED_CG_CONSTRAINED  : return CONVERGED_CG_CONSTRAINED;
         ^
5 warnings and 1 error generated.

**Comment by jwpeterson** [comment]
Looks like all warnings (which we should also fix) and the actual error is:
```
src/solvers/petsc_linear_solver.C:906:10: error: duplicate case value: 'KSP_CONVERGED_CG_CONSTRAINED' and 'KSP_CONVERGED_STEP_LENGTH' both equal '6'
case KSP_CONVERGED_STEP_LENGTH : return CONVERGED_STEP_LENGTH;
```

Since `KSP_CONVERGED_CG_CONSTRAINED` is apparently being deprecated, I'd just comment out that line (905) in your version of libmesh and then it should compile for you.

**Reply by aeronova** [reply]
Thank you for the tip. I commented out KSP_CONVERGED_CG_CONSTRAINED and the make process finished w/o errors.

**Comment by roystgnr** [comment]
Thanks for catching this.  Hopefully the errors are all fixed by #3517; that seems to be enough for my build configurations, at least.
